torch/distributed/_composable/fsdp/_fsdp_collectives.py

```diff
@@ -32,17 +38,22 @@ def foreach_all_gather(
     world_size, rank = group.size(), group.rank()
     # - Copy in
     with torch.cuda.stream(all_gather_copy_in_stream):
-        param_all_gather_inputs = [
-            fsdp_param.all_gather_input for fsdp_param in fsdp_params
+        param_all_gather_inputs: List[Tuple[torch.Tensor, ...]] = [
+            fsdp_param.all_gather_inputs for fsdp_param in fsdp_params
         ]
-        g = itertools.groupby(t.dtype for t in param_all_gather_inputs)
+        (
+            param_all_gather_input_dtypes,
+            param_all_gather_input_numels,
+        ) = _get_all_gather_input_metadatas(param_all_gather_inputs)
+        g = itertools.groupby(d for ds in param_all_gather_input_dtypes for d in ds)
         if next(g, True) and not next(g, False):  # same dtype
-            dtype = param_all_gather_inputs[0].dtype
+            dtype = param_all_gather_inputs[0][0].dtype
         else:
             raise NotImplementedError(
```

Comment:
This is coming in future PRs.

Reply:
Oh so we do allow mixing different dtypes, is this for NF4Tensor?

Reply:
This part is for mixing different dtypes, even across different parameters within one parameter group, so this includes the "fp8 all-gather" case where we mix fp8 and bf16 parameters.